framework/doc/content/source/csg/CSGSurface.md

```diff
@@ -0,0 +1,36 @@
+# CSGSurface
+
+This is an abstract class for any surface definition for [constructive solid geometry (CSG)](syntax/CSG/index.md) representations.
+The [source/csg/CSGBase.md] framework already contains definitions for some basic surfaces (planes, spheres, and cylinders), but developers of modules or MOOSE-based application can define additional custom surfaces using this abstract class.
+General information about implementing [!ac](CSG)-related methods can be found in [source/csg/CSGBase.md].
+
+## Defining a New Surface Type
+
+Any arbitrary or custom surface can be defined by inheriting from `CSGSurface`.
+At a minimum, the surface name, surface type, and boundary type, must be set, and two virtual methods that need to be defined.
+Additional information required by the constructor will depend on the specific surface to be defined.
+
+### Setting Coefficients
+
+Each surface should be defined by some equation, which therefore requires the definition of coefficients or other parameters.
+These coefficient values should be set and be returned through the `getCoefficients` method which will return a map of the coefficient strings to their values.
+These coefficient values are returned in the [!ac](JSON) file that is produced.
+For example, a plane is defined by the equation $ax + by + cz = d$ and so this method would return a map of the values for `a`, `b`, `c`, and `d`.
+
+### Half-space Determination
+
+In [!ac](CSG) representation, knowing which half-space of the surface is positive or negative is necessary for correct construction of a [cell](source/csg/CSGBase.md#cells) [region](source/csg/CSGBase.md#regions).
+To help determine the sign of these half-spaces, each surface type should have a `getHalfspaceFromPoint` method implemented that can be used to determine if a point lies within the positive (`CSGSurface::Direction::POSITIVE`) or negative (`CSGSurface::Direction::NEGATIVE`) half-space for the surface.
```

Comment:
Something we never addressed in our work is what should happen if the point lies on the equation on the surface?

Reply:
I'm thinking more about this, but what if we change getHalfspaceFromPoint to evaluateSurfaceEquationAtPoint that returns the evaulation of the surface at a specific point? This would be a virtual function, and then we define getHalfspaceFromPoint as:

```
auto eval = evaluateSurfaceEquationAtPoint(point);
if (eval > 0)
  return Direction::POSITIVE;
else if (eval < 0)
  return Direction::NEGATIVE;
else
  mooseError("Point lies on surface")
```

Reply:
We'd be using fuzzyLessThan or fuzzyGreaterThan instead of < and >

Reply:
I don't think this needs to return an error though. It is valid for a point to lie on a surface and whether that is considered positive or negative halfspace would depend on what the surface is specifically.

Reply:
Are you saying what halfspace a point is in is surface-dependent? I would think it would be ambiguous which halfspace such a point would be in, hence the error.

Reply:
ya the more I am thinking about it, the error does make sense. That wiki article sort of states that one of the halfspaces is considered the "equal-to" consideration. Just thinking about particle transport, it is possible for a particle to lie on a surface and the code will decide if that considers it to be in the pos or neg halfspace. But in reality when we are only looking at geometry construction, so we should only ever have clear division of space and if a point is exactly on a surface, that probably means there is a larger issue with the construction process.